src/repo/kv.rs

```diff
@@ -0,0 +1,373 @@
+use super::{Column, DataStore};
+use crate::error::Error;
+use crate::repo::{PinKind, PinMode, PinStore, References};
+use async_trait::async_trait;
+use cid::{self, Cid};
+use futures::stream::{StreamExt, TryStreamExt};
+use sled::{self, Config as DbConfig, Db, Mode as DbMode};
+use std::convert::Into;
+use std::path::PathBuf;
+use std::str::{self, FromStr};
+use tracing_futures::Instrument;
+
+#[derive(Debug)]
+pub struct KvDataStore {
+    path: PathBuf,
+    db: Option<Db>,
+}
+
+impl KvDataStore {
+    fn _put(&self, key: &str, value: &str) -> Result<(), Error> {
+        let db = self.get_db();
+
+        let _ = db.insert(key, value)?;
+
+        Ok(())
+    }
+
+    fn _remove(&self, key: &str) -> Result<(), Error> {
+        let db = self.get_db();
+
+        match db.remove(key) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn _apply_batch(&self, batch: sled::Batch) -> Result<(), Error> {
+        let db = self.get_db();
+
+        Ok(db.apply_batch(batch)?)
+    }
+
+    fn get_db(&self) -> &Db {
+        self.db.as_ref().unwrap()
+    }
+}
+
+#[async_trait]
+impl DataStore for KvDataStore {
+    fn new(root: PathBuf) -> KvDataStore {
+        KvDataStore {
+            path: root,
+            db: None,
+        }
+    }
+
+    async fn init(&self) -> Result<(), Error> {
+        let config = DbConfig::new();
+
+        let db = config
+            .mode(DbMode::HighThroughput)
+            .path(self.path.as_path())
+            .open()?;
+
+        unsafe {
+            let kv_ref = self as *const KvDataStore;
+            let kv_mut = kv_ref as *mut KvDataStore;
+            (*kv_mut).db = Some(db);
+        }
```

Comment:
Took me a while to decipher this but yeah ... `init(&self)` is an unfortunate wrinkle in the current design. Thanks for not changing this, this is much easier to review at least.. I cannot really see any workarounds for this at the moment even if this is quite unsafe.. Did you get any segfaults around this when testing? I wonder if there should at least be a memory barrier.

Reply:
Yes, `init(&self)` brings a trouble. I got no segfaults when testing.

Reply:
It might be good enough for now to use mutex and clone the Db as in https://github.com/fetchadd/rust-ipfs/pull/1